homeassistant/components/iotty/cover.py

```diff
@@ -0,0 +1,218 @@
+"""Implement a iotty Shutter Device."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any, cast
+
+from iottycloud.device import Device
+from iottycloud.shutter import Shutter, ShutterState
+from iottycloud.verbs import SH_DEVICE_TYPE_UID
+
+from homeassistant.components.cover import (
+    ATTR_POSITION,
+    CoverDeviceClass,
+    CoverEntity,
+    CoverEntityFeature,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import IottyConfigEntry
+from .api import IottyProxy
+from .const import DOMAIN
+from .coordinator import IottyDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: IottyConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Activate the iotty Shutter component."""
+    _LOGGER.debug("Setup COVER entry id is %s", config_entry.entry_id)
+
+    coordinator = config_entry.runtime_data.coordinator
+    entities = [
+        IottyShutter(
+            coordinator=coordinator, iotty_cloud=coordinator.iotty, iotty_device=d
+        )
+        for d in coordinator.data.devices
+        if d.device_type == SH_DEVICE_TYPE_UID
+        if (isinstance(d, Shutter))
+    ]
+    _LOGGER.debug("Found %d Shutters", len(entities))
+
+    async_add_entities(entities)
+
+    known_devices: set = config_entry.runtime_data.known_devices
+    for known_device in coordinator.data.devices:
+        if known_device.device_type == SH_DEVICE_TYPE_UID:
+            known_devices.add(known_device)
+
+    @callback
+    def async_update_data() -> None:
+        """Handle updated data from the API endpoint."""
+        if not coordinator.last_update_success:
+            return
+
+        devices = coordinator.data.devices
+        entities = []
+        known_devices: set = config_entry.runtime_data.known_devices
+
+        # Add entities for devices which we've not yet seen
+        for device in devices:
+            if (
+                any(d.device_id == device.device_id for d in known_devices)
+                or device.device_type != SH_DEVICE_TYPE_UID
+            ):
+                continue
+
+            iotty_entity = IottyShutter(
+                coordinator=coordinator,
+                iotty_cloud=coordinator.iotty,
+                iotty_device=Shutter(
+                    device.device_id,
+                    device.serial_number,
+                    device.device_type,
+                    device.device_name,
+                ),
+            )
+
+            entities.extend([iotty_entity])
+            known_devices.add(device)
```

Comment:
This looks really inefficient. can't you keep a list in the cover platform with all the known shutters and just watch out for them?

Reply:
Hi, it's not clear to me what you mean with 

> keep a list in the cover platform with all the known shutters and just watch out for them

Moreover, this approach is already used in the Switch entity as well, and we didn't have any problems with it. I understand there might be a bit of extra calculations going on here, but there doesn't seem to be any concern in efficiency for the size of the lists involved here